compiler/optimizer/OMRSimplifierHandlers.cpp

```diff
@@ -12845,6 +12845,151 @@ TR::Node *su2dSimplifier(TR::Node * node, TR::Block * block, TR::Simplifier * s)
    }
 
 
+bool canTransformOp(bool canTransformAdd, bool canTransformSub, TR::Node *node, TR::Simplifier *s)
+   {
+   if (!(canTransformAdd || canTransformSub))
+      {
+      if (s->trace())
+         traceMsg(s->comp(),
+                  "\nEliminating add/sub under order compare node n%dn failed due to overflow\n",
+                  node->getGlobalIndex());
+
+      return false;
+      }
+   return true;
+   }
+
+TR::Node *
+arithUnderSignedAndUnsignedCompareHandler(TR::Node *opNode, TR::Node *secondChild, TR::Simplifier *s, TR::Node *node,
+                                          bool isSigned, bool isAddOp, bool isSubOp)
+   {
+
+   TR::DataType DataType = secondChild->getDataType();
+   if (!isSigned)
+      {
+      uint64_t oldUConst1 = 0, oldUConst2 = 0, unsignedMax = 0;
+      switch (DataType)
+         {
+         case TR::Int8:
+            oldUConst1 = opNode->getSecondChild()->getUnsignedByte();
+            oldUConst2 = secondChild->getUnsignedByte();
+            unsignedMax = TR::getMaxUnsigned<TR::Int8>();
+            break;
+         case TR::Int16:
+            oldUConst1 = opNode->getSecondChild()->getUnsignedShortInt();
+            oldUConst2 = secondChild->getUnsignedShortInt();
+            unsignedMax = TR::getMaxUnsigned<TR::Int16>();
+            break;
+         case TR::Int32:
+            oldUConst1 = opNode->getSecondChild()->getUnsignedInt();
+            oldUConst2 = secondChild->getUnsignedInt();
+            unsignedMax = TR::getMaxUnsigned<TR::Int32>();
+            break;
+         case TR::Int64:
+            oldUConst1 = opNode->getSecondChild()->getUnsignedLongInt();
+            oldUConst2 = secondChild->getUnsignedLongInt();
+            unsignedMax = TR::getMaxUnsigned<TR::Int64>();
+            break;
+         default:
+            if (s->trace())
+               traceMsg(s->comp(),
+                        "\nEliminating add/sub under compare node n%dn failed due to opcode data type\n",
+                        node->getGlobalIndex());
+
+            TR_ASSERT_FATAL(false, "\nEliminating add/sub under compare node n%dn failed due to opcode data type\n",
+                            node->getGlobalIndex());
+            break;
+         }
+      //under/overflow handling:
+      //Must detect under/overflow without causing under/overflow
+      //Transformation x-a>b ----> x>b+a has two cases of under/overflow:
+      //overflow of x-a and overflow of b+a 
+      //x-a is handled by cannotOverflow flag;  only concerned about under/overflow from constant transformation. 
+      
+     
+      //For Unsigned Operations, a, b are positive, so add op must handle underFlow
+      //x+a>b  --opt--> x>b-a CASE OF UNDERFLOW IF b-a<unsignedMin
+      //unsignedMin is always 0
```

Comment:
It wasn't obvious to me on a first read that these `a` refers to `oldUConst1` and that `b` refers to `oldUConst2`. Might there be a way of making this clearer?